code/datums/components/bayonet_charge.dm

```diff
@@ -0,0 +1,183 @@
+	///The list of melee/not gun weapons available to be charged with
+	var/list/enabled_melee_weapons = list(
+		/obj/item/weapon/powerfist,
+		/obj/item/weapon/claymore/mercsword/machete,
+		/obj/item/weapon/claymore/mercsword/officersword,
+		/obj/item/weapon/twohanded/spear/tactical,
+	)
+
+/datum/component/bayonetcharge/RemoveComponent()
+	. = ..()
+	child.remove_action(charger)
+
+/datum/component/bayonetcharge/Initialize(...)
+	. = ..()
+	weaponinhand = src.parent
+	charger = weaponinhand.gun_user
+	child = new
+	child.give_action(src.charger)
+	child.parent = src
+
+/datum/action/bayonetcharge
+	name = "Toggle Bayonet Charge"
+	action_icon_state = "charge_human"
+
+	var/datum/component/bayonetcharge/parent
+
+/datum/action/bayonetcharge/give_action(mob/M)
+	. = ..()
+
+/datum/action/bayonetcharge/remove_action(mob/M)
+
+
+	parent.charge_off()
+	. = ..()
+
+/datum/action/bayonetcharge/action_activate()
+	if(parent.amcharging)
+		return
+	if(parent.chargewhenmove == TRUE)
+		parent.charge_off()
+		return
+
+	parent.charge_on()
+
+/datum/component/bayonetcharge/proc/charge_on()
+	if(incooldown)
+		to_chat(charger, "<span class='warning'>Charge is still on cooldown!</span>")
+		return
+	chargewhenmove = TRUE
+	child.add_selected_frame()
+	 //just in case you didnt activate the action, charger would be null if you do.
+	to_chat(charger, "<span class='warning'>You stand ready to charge</span>")
+	RegisterSignal(charger, COMSIG_MOVABLE_MOVED, .proc/update_charging)
+	RegisterSignal(charger, COMSIG_ATOM_DIR_CHANGE, .proc/on_dir_change)
+
+/datum/component/bayonetcharge/proc/charge_off()
+	chargewhenmove = FALSE
+
+	to_chat(charger, "<span class='warning'>You no longer stand ready to charge</span>")
+	if(charger) //locks up if you dont check if charger is existant
+		UnregisterSignal(charger, COMSIG_MOVABLE_MOVED)
+		UnregisterSignal(charger, COMSIG_ATOM_DIR_CHANGE)
+		UnregisterSignal(charger, list(COMSIG_MOVABLE_PREBUMP_TURF, COMSIG_MOVABLE_PREBUMP_MOVABLE, COMSIG_MOVABLE_PREBUMP_EXIT_MOVABLE))
+	child.remove_selected_frame()
+	stop_charge()
+
+
+/datum/component/bayonetcharge/proc/on_dir_change(datum/source, old_dir, new_dir)
+	SIGNAL_HANDLER
+	if(old_dir == new_dir)
+		return
+	stop_charge()
+//called when the marine moves
+/datum/component/bayonetcharge/proc/update_charging(datum/source, atom/oldloc, direction, Forced)
+	SIGNAL_HANDLER_DOES_SLEEP
+	if(movementcontinuationlimit && world.time > movementcontinuationlimit) //if you have stopped moving, it resetss the steps taken coutner
+		if(amcharging)
+			amcharging = FALSE
+			incooldown = TRUE
+			addtimer(CALLBACK(src, .proc/cooldownend), cooldowntime)
+		stop_charge()
+		movementcontinuationlimit = null
+		return
+	stepstaken++
+	movementcontinuationlimit = world.time + 10
+
+	handle_charge()
+
+/datum/component/bayonetcharge/proc/handle_charge()
+
+	//actual charging is under here
+	var/chargespeed = clamp(stepstaken * speedperstep, 0, maxspeed)
+	if(charger.incapacitated())
```

Comment:
the way you do it to not have a charger var, is by casting parent into a mob in procs.

so here, you would do var/mob/living/charger = parent in each proc where you'd need charger